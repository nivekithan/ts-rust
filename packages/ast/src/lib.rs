pub mod data_type;
pub mod declaration;
pub mod expression;

use data_type::DataType;
use declaration::{Declaration, VariableAssignmentOperator, VariableDeclarationKind};
use expression::{BinaryOperator, Expression, UnaryOperator};

#[derive(Debug, PartialEq, Clone)]
pub enum Ast {
    Expression(Expression),
    Declaration(Declaration),
}

// constructors

impl Ast {
    pub fn new_float_literal(name: &String, value: f64) -> Ast {
        return Ast::Expression(Expression::FloatLiteralExp {
            name: name.to_string(),
            value,
        });
    }

    pub fn new_boolean_literal(name: &String, value: bool) -> Ast {
        return Ast::Expression(Expression::BooleanLiteralExp {
            name: name.to_string(),
            value,
        });
    }

    pub fn new_string_literal(name: &String) -> Ast {
        return Ast::Expression(Expression::StringLiteralExp {
            value: name.to_string(),
        });
    }

    pub fn new_unary_exp(argument: Box<Expression>, operator: UnaryOperator) -> Ast {
        return Ast::Expression(Expression::UnaryExp { operator, argument });
    }

    pub fn new_binary_exp(
        left: Box<Expression>,
        right: Box<Expression>,
        operator: BinaryOperator,
    ) -> Ast {
        return Ast::Expression(Expression::BinaryExp {
            operator,
            left,
            right,
        });
    }
    pub fn new_variable_declaration(
        ident_name: &str,
        exp: Expression,
        kind: VariableDeclarationKind,
    ) -> Ast {
        return Ast::Declaration(Declaration::VariableDeclaration {
            ident_name: ident_name.to_string(),
            exp,
            kind,
        });
    }

    pub fn new_variable_assignment(
        ident_name: &str,
        operator: VariableAssignmentOperator,
        exp: Expression,
    ) -> Ast {
        return Ast::Declaration(Declaration::VariableAssignment {
            operator,
            ident_name: ident_name.to_string(),
            exp,
        });
    }

    pub fn new_if_block(condition: Expression, blocks: Vec<Ast>) -> Ast {
        if let DataType::Boolean = condition.get_data_type() {
            return Ast::Declaration(Declaration::IfBlockDeclaration {
                condition,
                block: Box::new(blocks),
            });
        } else {
            panic!("Condition can only be expression whose datatype is boolean")
        }
    }
}

impl Ast {
    pub fn get_data_type(&self) -> Result<DataType, String> {
        match self {
            Ast::Expression(exp) => {
                return Ok(exp.get_data_type());
            }
            Ast::Declaration(dec) => {
                return Err(format!(
                    "There is no datatype associated with declaration {:?}",
                    dec
                ));
            }
        }
    }
}
