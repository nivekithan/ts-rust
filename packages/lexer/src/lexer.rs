use crate::token::LiteralKind;

use super::token::KeywordKind;
use core::panic;
use std::str::Chars;

use super::token::Token;

pub(crate) struct Lexer<'a> {
    content: Chars<'a>,
    cur_char: Option<char>,
}

impl<'a> Lexer<'a> {
    pub(crate) fn new(input: &'a str) -> Lexer<'a> {
        let content = input.chars();

        let mut lexer = Lexer {
            content,
            cur_char: None,
        };

        lexer.next();

        return lexer;
    }

    pub(crate) fn next_token(&mut self) -> Token {
        self.eat_whitespace();

        let cur_char = self.cur_char;

        match cur_char {
            None => Token::Eof,
            Some(char) => {
                use Token::*;

                if char == ';' {
                    self.next(); // consumes ;
                    return SemiColon;
                } else if char == ':' {
                    self.next(); // consumes :
                    return Colon;
                } else if char == '=' {
                    self.next(); // consumes =
                    return Assign;
                } else if char == '{' {
                    self.next(); // consumes {
                    return AngleOpenBracket;
                } else if char == '}' {
                    self.next(); // consumes }
                    return AngleCloseBracket;
                } else if char == '(' {
                    self.next(); // consumes (
                    return CurveOpenBracket;
                } else if char == ')' {
                    self.next(); // consumes )
                    return CurveCloseBracket;
                } else if char == ',' {
                    self.next(); // consumes ,
                    return Comma;
                } else if char == '!' {
                    self.next(); // consumes !
                    return Bang;
                } else if char == '+' {
                    self.next(); // consumes +
                    return Plus;
                } else if char == '-' {
                    self.next(); // consumes -
                    return Minus;
                } else if char == '*' {
                    self.next(); // consumes +
                    return Star;
                } else if char == '/' {
                    self.next(); // consumes /
                    return Slash;
                } else if char == '|' {
                    self.next(); // consumes |
                    return VerticalBar;
                } else if char == '^' {
                    self.next(); // consumes ^
                    return Caret;
                } else if char == '&' {
                    self.next(); // consumes &
                    return Ampersand;
                } else if char == '\'' {
                    let string_name = self.read_string('\'');
                    return Literal(LiteralKind::String { name: string_name });
                } else if char == '\"' {
                    let string_name = self.read_string('\"');
                    return Literal(LiteralKind::String { name: string_name });
                } else if char == '`' {
                    let string_name = self.read_string('`');
                    return Literal(LiteralKind::String { name: string_name });
                } else {
                    if is_letter(&char) {
                        let ident_name = self.read_identifier();
                        let may_be_keyword = is_keyword(&ident_name);

                        match may_be_keyword {
                            IsKeyword::Yes(kind) => return Keyword(kind),
                            IsKeyword::No => return Ident { name: ident_name },
                        }
                    } else if is_digit(&char) {
                        let digit_name = self.read_digit();
                        let digit_value: f64 = digit_name.parse().unwrap();

                        return Literal(LiteralKind::Float {
                            name: digit_name,
                            value: digit_value,
                        });
                    }
                    self.next(); // consumes Illegal
                    return Illegal;
                }
            }
        }
    }

    fn next(&mut self) -> Option<char> {
        let next_cur = self.content.next();
        self.cur_char = next_cur;

        return next_cur;
    }

    // If cur_char is whitespace it wil eat char until cur_char is not whitespace
    fn eat_whitespace(&mut self) {
        let cur_char = self.cur_char;

        if let Some(mut c) = cur_char {
            while is_whitespace(&c) {
                let next_char = self.next();
                if let Some(ch) = next_char {
                    c = ch
                } else {
                    break;
                }
            }
        }
    }

    // After executing this function the cur_char wont be a letter
    fn read_identifier(&mut self) -> String {
        let mut ident_name = self.cur_char.expect("unreachable").to_string();
        let next_cur = self.next();

        match next_cur {
            None => return ident_name,
            Some(mut c) => {
                while is_letter(&c) {
                    ident_name.push(c);

                    let next_cur = self.next();

                    match next_cur {
                        None => return ident_name,
                        Some(ch) => c = ch,
                    }
                }

                return ident_name;
            }
        }
    }
    // After executing this function the cur_char wont be a digit
    fn read_digit(&mut self) -> String {
        let mut digit_name = self.cur_char.expect("unreachable").to_string();

        let next_cur = self.next();

        match next_cur {
            None => return digit_name,
            Some(mut c) => {
                while is_digit(&c) {
                    digit_name.push(c);

                    let next_cur = self.next();

                    match next_cur {
                        None => return digit_name,
                        Some(ch) => c = ch,
                    }
                }

                return digit_name;
            }
        }
    }
    // Assumes the cur_char is starting char of string_literal
    // Ex: ' " `
    // It will end after consuming the end char that
    // cur_char wont be end_char
    fn read_string(&mut self, end_char: char) -> String {
        let mut string_name = String::new();
        loop {
            let next_char = self.next();

            match next_char {
                None => panic!("Lexer error expected {} before end of file", end_char),
                Some(ch) => {
                    if ch == end_char {
                        self.next(); // consumes end_char
                        break;
                    } else {
                        string_name.push(ch)
                    }
                }
            }
        }

        return string_name;
    }
}

// From rust source code
fn is_whitespace(c: &char) -> bool {
    return matches!(
        c,
        '\u{0009}'   // \t
        | '\u{000A}' // \n
        | '\u{000B}' // vertical tab
        | '\u{000C}' // form feed
        | '\u{000D}' // \r
        | '\u{0020}' // space

        // NEXT LINE from latin1
        | '\u{0085}'

        // Bidi markers
        | '\u{200E}' // LEFT-TO-RIGHT MARK
        | '\u{200F}' // RIGHT-TO-LEFT MARK

        // Dedicated whitespace characters from Unicode
        | '\u{2028}' // LINE SEPARATOR
        | '\u{2029}' // PARAGRAPH SEPARATOR)
    );
}

fn is_letter(c: &char) -> bool {
    return matches!(
        c,
         'a'..='z'
        | 'A'..='Z'
        | '_'
    );
}

fn is_digit(c: &char) -> bool {
    return matches!(c, '0'..='9');
}

pub enum IsKeyword {
    Yes(KeywordKind),
    No,
}

fn is_keyword(word: &String) -> IsKeyword {
    if word == "const" {
        return IsKeyword::Yes(KeywordKind::Const);
    } else if word == "true" {
        return IsKeyword::Yes(KeywordKind::True);
    } else if word == "false" {
        return IsKeyword::Yes(KeywordKind::False);
    } else if word == "let" {
        return IsKeyword::Yes(KeywordKind::Let);
    } else if word == "if" {
        return IsKeyword::Yes(KeywordKind::If);
    } else {
        return IsKeyword::No;
    }
}
