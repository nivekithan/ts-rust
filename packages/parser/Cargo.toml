[package]
name = "parser"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lexer = { path = "../lexer"}
ast = {path = "../ast"}
indexmap = "1.7.0"
