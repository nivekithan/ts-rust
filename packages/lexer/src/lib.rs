mod lexer;
mod token;

use crate::lexer::Lexer;
use crate::token::Token;

#[cfg(test)]
mod test;

pub fn convert_to_token(input: &str) -> Vec<Token> {
    let mut lexer = Lexer::new(input);
    let mut tokens: Vec<Token> = Vec::new();

    loop {
        let token = lexer.next_token();

        tokens.push(token);

        if let Token::Eof = token {
            break;
        }
    }

    return tokens;
}
