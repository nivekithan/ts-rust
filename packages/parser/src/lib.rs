mod parse_block;
mod parse_expression;
mod parse_type_expression;
mod parser;
mod symbol_table;
mod utils;

#[cfg(test)]
mod tests;

use crate::{parser::Parser, symbol_table::SymbolContext};
use ast::Ast;
use lexer::token::Token;

pub fn convert_to_ast(input: Vec<Token>) -> Vec<Ast> {
    let mut parser = Parser::new(&input);
    let mut asts: Vec<Ast> = vec![];
    let mut context = SymbolContext::new_empty_context();

    while parser.get_cur_token().unwrap() != &Token::Eof {
        let next_ast = parser.next_ast(&mut context);
        asts.push(next_ast);
    }

    return asts;
}

#[cfg(test)]
mod test {
    use std::collections::HashMap;

    use ast::{
        data_type::DataType,
        declaration::{VariableDeclarationKind},
        expression::Expression,
        Ast,
    };
    use indexmap::indexmap;
    use lexer::convert_to_token;

    use crate::convert_to_ast;

    #[test]
    fn test_2() {
        let input = "
        const x = {a : 1, b : 2};
        const y = x.a;";

        let mut exp_hash_map: HashMap<String, Expression> = HashMap::new();

        exp_hash_map.insert(
            "a".to_string(),
            Expression::FloatLiteralExp {
                name: "1".to_string(),
                value: 1.0,
            },
        );
        exp_hash_map.insert(
            "b".to_string(),
            Expression::FloatLiteralExp {
                name: "2".to_string(),
                value: 2.0,
            },
        );

        let expected_output: Vec<Ast> = vec![
            Ast::new_variable_declaration(
                "x_",
                Expression::ObjectLiteral {
                    expression: exp_hash_map,
                    data_type: DataType::ObjectType {
                        entries: indexmap! {"a".to_string() => DataType::Float, "b".to_string() => DataType::Float},
                    },
                },
                VariableDeclarationKind::Const,
            ),
            Ast::new_variable_declaration(
                "y_",
                Expression::DotMemberAccess {
                    container: Box::new(Expression::IdentExp {
                        data_type: DataType::ObjectType {
                            entries: indexmap! {"a".to_string() => DataType::Float, "b".to_string() => DataType::Float},
                        },
                        name: "x_".to_string(),
                    }),
                    argument: "a".to_string(),
                },
                VariableDeclarationKind::Const,
            ),
        ];

        let actual_output = convert_to_ast(convert_to_token(input));

        assert_eq!(expected_output, actual_output);
    }

    #[test]
    fn test_3() {
        let input = "
        function foo(x : number) : number {
            const x = 1;
        }";

        let expected_output = vec![Ast::new_function_declaration(
            indexmap! {"x".to_string() => DataType::Float},
            Box::new(vec![Ast::new_variable_declaration(
                "x_",
                Expression::FloatLiteralExp {
                    name: "1".to_string(),
                    value: 1.0,
                },
                VariableDeclarationKind::Const,
            )]),
            "foo_".to_string(),
            DataType::Float,
        )];

        let actual_output = convert_to_ast(convert_to_token(input));

        assert_eq!(expected_output, actual_output);
    }
}
