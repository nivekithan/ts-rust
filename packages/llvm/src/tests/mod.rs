use ast::{
    data_type::DataType,
    expression::{BinaryOperator, Expression},
};
use lexer::convert_to_token;
use parser::convert_to_ast;
use test_utils::{generate_expressions, ExpressionForm};

use crate::write_llvm_ir;

#[test]
fn test_float_literal_exp() {
    let input = "
    const x = 1
    ";

    let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input)));

    insta::assert_snapshot!(actual_output, @r###"
    ; ModuleID = 'main'
    source_filename = "main"

    define void @main() {
    entry:
      %x = alloca double, align 8
      store double 1.000000e+00, double* %x, align 8
      ret void
    }
    "###);
}

// #[test]
// fn test_string_literal_exp() {
//     let input = "
//     const x = \"1\";";

//     let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input)));

//     insta::assert_snapshot!(actual_output, @"");
// }

#[test]
fn test_boolean_literal_true_exp() {
    let input = "
    const x = true";

    let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input)));

    insta::assert_snapshot!(actual_output, @r###"
    ; ModuleID = 'main'
    source_filename = "main"

    define void @main() {
    entry:
      %x = alloca i64, align 8
      store i64 1, i64* %x, align 4
      ret void
    }
    "###);
}

#[test]
fn test_boolean_literal_false_exp() {
    let input = "
    const x = false";

    let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input)));

    insta::assert_snapshot!(actual_output, @r###"
    ; ModuleID = 'main'
    source_filename = "main"

    define void @main() {
    entry:
      %x = alloca i64, align 8
      store i64 0, i64* %x, align 4
      ret void
    }
    "###);
}

#[test]
fn test_float_ident_exp() {
    let input = "
    const x = 1;
    const y = x";

    let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input)));

    insta::assert_snapshot!(actual_output, @r###"
    ; ModuleID = 'main'
    source_filename = "main"

    define void @main() {
    entry:
      %x = alloca double, align 8
      store double 1.000000e+00, double* %x, align 8
      %y = alloca double, align 8
      %"1" = load double, double* %x, align 8
      store double %"1", double* %y, align 8
      ret void
    }
    "###);
}

#[test]
fn test_bool_ident_exp() {
    let input = "
    const x = true;
    const y = x";

    let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input)));

    insta::assert_snapshot!(actual_output, @r###"
    ; ModuleID = 'main'
    source_filename = "main"

    define void @main() {
    entry:
      %x = alloca i64, align 8
      store i64 1, i64* %x, align 4
      %y = alloca i64, align 8
      %"1" = load i64, i64* %x, align 4
      store i64 %"1", i64* %y, align 4
      ret void
    }
    "###);
}

// #[test]
// fn test_string_ident_exp() {
//     let input = "
//     const x = \"1\";
//     const y = x";

//     let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input)));

//     insta::assert_snapshot!(actual_output, @"");
// }

#[test]
fn test_unary_plus_exp() {
    let expression_forms = generate_expressions(&DataType::Float, "_x");

    for exp_form in expression_forms {
        if is_working_in_progress(&exp_form) {
            continue;
        }

        println!("{:?}", exp_form);

        let input = exp_form.generate_input(format!("const x = +({})", exp_form.main_string));

        let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

        insta::assert_snapshot!(input, actual_output);
    }
}

#[test]
fn test_unary_minus_exp() {
    let expression_forms = generate_expressions(&DataType::Float, "_x");

    for exp_form in expression_forms {
        if is_working_in_progress(&exp_form) {
            continue;
        }

        let input = exp_form.generate_input(format!("const x = -({})", exp_form.main_string));

        let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

        insta::assert_snapshot!(input, actual_output);
    }
}

#[test]
fn test_unary_bang_exp() {
    let expression_forms = generate_expressions(&DataType::Boolean, "_x");

    for exp_form in expression_forms {
        if is_working_in_progress(&exp_form) {
            continue;
        }

        let input = exp_form.generate_input(format!("const x = !({})", exp_form.main_string));

        let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

        insta::assert_snapshot!(input, actual_output);
    }
}

#[test]
fn test_binary_plus_float_exp() {
    let expression_forms_1 = generate_expressions(&DataType::Float, "_x");

    for exp_form_1 in expression_forms_1 {
        let expression_forms_2 = generate_expressions(&DataType::Float, "_y");

        if is_working_in_progress(&exp_form_1) {
            continue;
        }

        for exp_form_2 in expression_forms_2 {

            if  is_working_in_progress(&exp_form_2) {
                continue;
            }

            let input = exp_form_1.generate_input(exp_form_2.generate_input(format!(
                "
            const x = ({}) + ({})",
                exp_form_1.main_string, exp_form_2.main_string
            )));

            let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, actual_output);
        }
    }
}

// #[test]
// fn test_binary_plus_string_exp() {
//     let expression_forms_1 = generate_expressions(&DataType::String, "_x");

//     for exp_form_1 in expression_forms_1 {
//         let expression_forms_2 = generate_expressions(&DataType::String, "_y");

//         for exp_form_2 in expression_forms_2 {

//             if is_working_in_progress(&exp_form_1) || is_working_in_progress(&exp_form_2) {
//                 continue;
//             }

//             let input = exp_form_1.generate_input(exp_form_2.generate_input(format!(
//                 "
//             const x = ({}) + ({})",
//                 exp_form_1.main_string, exp_form_2.main_string
//             )));

//             let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

//             insta::assert_snapshot!(actual_output, @"");
//         }
//     }
// }

#[test]
fn test_binary_minus_exp() {
    let expression_forms_1 = generate_expressions(&DataType::Float, "_x");

    for exp_form_1 in expression_forms_1 {
        let expression_forms_2 = generate_expressions(&DataType::Float, "_y");

        if is_working_in_progress(&exp_form_1) {
            continue;
        }

        for exp_form_2 in expression_forms_2 {

            if  is_working_in_progress(&exp_form_2) {
                continue;
            }

            let input = exp_form_1.generate_input(exp_form_2.generate_input(format!(
                "
            const x = ({}) - ({})",
                exp_form_1.main_string, exp_form_2.main_string
            )));

            let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, actual_output);
        }
    }
}

#[test]
fn test_binary_star_exp() {
    let expression_forms_1 = generate_expressions(&DataType::Float, "_x");

    for exp_form_1 in expression_forms_1 {
        let expression_forms_2 = generate_expressions(&DataType::Float, "_y");

        if is_working_in_progress(&exp_form_1) {
            continue;
        }

        for exp_form_2 in expression_forms_2 {

            if  is_working_in_progress(&exp_form_2) {
                continue;
            }

            let input = exp_form_1.generate_input(exp_form_2.generate_input(format!(
                "
            const x = ({}) * ({})",
                exp_form_1.main_string, exp_form_2.main_string
            )));

            let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, actual_output);
        }
    }
}

#[test]
fn test_binary_slash_exp() {
    let expression_forms_1 = generate_expressions(&DataType::Float, "_x");

    for exp_form_1 in expression_forms_1 {
        let expression_forms_2 = generate_expressions(&DataType::Float, "_y");

        if is_working_in_progress(&exp_form_1) {
            continue;
        }

        for exp_form_2 in expression_forms_2 {

            if  is_working_in_progress(&exp_form_2) {
                continue;
            }

            let input = exp_form_1.generate_input(exp_form_2.generate_input(format!(
                "
            const x = ({}) / ({})",
                exp_form_1.main_string, exp_form_2.main_string
            )));

            let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, actual_output);
        }
    }
}

// #[test]
// fn test_binary_vertical_bar_exp() {
//     let expression_forms_1 = generate_expressions(&DataType::Float, "_x");

//     for exp_form_1 in expression_forms_1 {
//         let expression_forms_2 = generate_expressions(&DataType::Float, "_y");

//         for exp_form_2 in expression_forms_2 {

//             if is_working_in_progress(&exp_form_1) || is_working_in_progress(&exp_form_2) {
//                 continue;
//             }

//             let input = exp_form_1.generate_input(exp_form_2.generate_input(format!(
//                 "
//             const x = ({}) | ({})",
//                 exp_form_1.main_string, exp_form_2.main_string
//             )));

//             let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

//             insta::assert_snapshot!(actual_output, @"");
//         }
//     }
// }

// #[test]
// fn test_binary_caret_exp() {
//     let expression_forms_1 = generate_expressions(&DataType::Float, "_x");

//     for exp_form_1 in expression_forms_1 {
//         let expression_forms_2 = generate_expressions(&DataType::Float, "_y");

//         for exp_form_2 in expression_forms_2 {

//             if is_working_in_progress(&exp_form_1) || is_working_in_progress(&exp_form_2) {
//                 continue;
//             }

//             let input = exp_form_1.generate_input(exp_form_2.generate_input(format!(
//                 "
//             const x = ({}) ^ ({})",
//                 exp_form_1.main_string, exp_form_2.main_string
//             )));

//             let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

//             insta::assert_snapshot!(actual_output, @"");
//         }
//     }
// }

// #[test]
// fn test_binary_ampersand_exp() {
//     let expression_forms_1 = generate_expressions(&DataType::Float, "_x");

//     for exp_form_1 in expression_forms_1 {
//         let expression_forms_2 = generate_expressions(&DataType::Float, "_y");

//         for exp_form_2 in expression_forms_2 {

//             if is_working_in_progress(&exp_form_1) || is_working_in_progress(&exp_form_2) {
//                 continue;
//             }

//             let input = exp_form_1.generate_input(exp_form_2.generate_input(format!(
//                 "
//             const x = ({}) & ({})",
//                 exp_form_1.main_string, exp_form_2.main_string
//             )));

//             let actual_output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

//             insta::assert_snapshot!(actual_output, @"");
//         }
//     }
// }

fn is_working_in_progress(exp_form: &ExpressionForm) -> bool {
    let final_expression = &exp_form.main_exp;

    let result = match final_expression {
        Expression::BinaryExp {
            operator,
            left: _,
            right: _,
        } => match operator {
            BinaryOperator::Ampersand | BinaryOperator::Caret | BinaryOperator::VerticalBar => true,
            _ =>  false,
        },

        Expression::UnaryExp {
            operator: _,
            argument: _,
        } => return false,

        Expression::FloatLiteralExp { name: _, value: _ } => return true,
        Expression::BooleanLiteralExp { name: _, value: _ } => return false,

        Expression::StringLiteralExp { value: _ } => true,

        Expression::IdentExp { name: _, data_type } => match data_type {
            DataType::String => true,

            _ => false,
        },
    };

    if !result {
        println!("Skipping expression {:?}", final_expression);
    } else {
        println!("Accepting expression {:?}", final_expression);
    }

    return result;
}
