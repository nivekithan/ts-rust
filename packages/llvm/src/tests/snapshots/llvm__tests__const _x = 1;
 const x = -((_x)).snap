---
source: packages/llvm/src/tests/mod.rs
expression: actual_output

---
; ModuleID = 'main'
source_filename = "main"

define void @main() {
entry:
  %_x = alloca double, align 8
  store double 1.000000e+00, double* %_x, align 8
  %x = alloca double, align 8
  %"2" = load double, double* %_x, align 8
  %"1" = fneg double %"2"
  store double %"1", double* %x, align 8
  ret void
}

